catlearn/ga/algorithm.py

```diff
@@ -237,42 +247,47 @@ def _get_fitness(self, param_list):
             The fitness based on the new parameters.
         """
         # Initialize array.
-        fit = np.zeros((len(param_list), self.fitness_parameters))
-
-        bool_list = np.asarray(param_list, dtype=np.bool)
-        for index in trange(bool_list.shape[0], leave=False,
-                            desc='working generation {}'.format(self.step + 1)
-                            ):
-            parameter = bool_list[index]
-            calc_fit = np.zeros(self.fitness_parameters)
-            for k in range(self.nsplit):
-                # Sort out training and testing data.
-                train_features = copy.deepcopy(self.features)
-                train_targets = copy.deepcopy(self.targets)
-                test_features = train_features.pop(k)[:, parameter]
-                test_targets = train_targets.pop(k)
-
-                train_features = np.concatenate(train_features,
-                                                axis=0)[:, parameter]
-                train_targets = np.concatenate(train_targets, axis=0)
-                try:
-                    calc_fit += np.array(self.fit_func(
-                        train_features, train_targets, test_features,
-                        test_targets))
-                except ValueError:
-                    # If there is a problem calculating fitness assign -inf.
-                    calc_fit += np.array(
-                        [float('-inf')] * self.fitness_parameters)
-                    msg = 'The fitness function is failing. Returning -inf.'
-                    warnings.warn(msg)
-
-            fit[index] = calc_fit / float(self.nsplit)
+        if self.nprocs == 1:
+            fit = np.zeros((len(param_list), self.fitness_parameters))
+
+            bool_list = np.asarray(param_list, dtype=np.bool)
+            for index in trange(bool_list.shape[0], leave=False,
+                                desc='working generation {}'.format(self.step +
+                                                                    1)):
+                args = (index, bool_list, self.fitness_parameters,
+                        self.features,
+                        self.targets,
+                        self.fit_func,
+                        self.nsplit)
+                calc_fit = _cross_validate(args)[1]
+                fit[index] = calc_fit / float(self.nsplit)
+        else:
+            fit = self._parallel_iterator(param_list)
 
         if self.pareto:
             fit = self._pareto_trainsform(fit)
 
         return np.reshape(fit, (len(fit),))
 
+    def _parallel_iterator(self, param_list):
```

Comment:
At this point it would probably be really good if we could sort the `param_list`. I think it is safe to assume that training a model with more features will be slower than training on with fewer. Therefor we really want to submit the models with more features first so the parallelization is less likely to be waiting for a slow model at the end holding everything up?

Reply:
Also, can you add docstrings?